File: tlg/src/main/java/ch/bailu/tlg/Square.java
package ch.bailu.tlg;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;


public class Square {
    private final static int BACKGROUND=0;
    private final static int COLOR=1;
    private final static int GREYED=2;
    private int mode=BACKGROUND;
    private int color=0;

    public TlgRectangle rect = new TlgRectangle(0,0,0,0);
    
    public Square () {}
    public Square (Square s) {
        set(s);
    }

    public Square(BufferedInputStream reader) throws IOException {
        color = ByteInteger.read(reader);
        mode=reader.read();
    }

    public void writeState(BufferedOutputStream output) throws IOException {
        ByteInteger.wrap(color).writeState(output);
        output.write(mode);
    }
    
    public boolean isVisible() {
        return (mode != BACKGROUND);
    }
    public boolean isInvisible() {
        return !isVisible();
    }

    public void set(Square s) {
        color=s.color;
        mode=s.mode;
    }

    public void setColor(int c) {
        color = c;
    }
    
    public boolean isActivated () {
        return isVisible(); 
    }
    
    public void makeVisible() {
        mode = COLOR;
    }
    
    public void makeInvisible() {
        mode = BACKGROUND;
    }

    public boolean isGreyedOut() {
        return (mode == GREYED);
    }
    
    public void enableGreyedOut() {
        if (isInvisible()) mode = GREYED;
    }
    
    public void disableGreyedOut() {
        if (isGreyedOut()) makeInvisible();
    }

    private void drawGrid(PlatformContext gc) {
        gc.drawFilledRectangle(gc.colorBackground(), rect);
        gc.drawLine(gc.colorGrid(), rect.getTL(), rect.getTR()); 
        gc.drawLine(gc.colorGrid(), rect.getTL(), rect.getBL());
    }
    private void draw3D (PlatformContext gc, int c) {
        TlgRectangle fillRect=new TlgRectangle(rect);
        fillRect.shrink(1);
        gc.drawFilledRectangle(c, fillRect);
        
        gc.drawLine(gc.colorDark(), rect.getTL(), rect.getTR());
        gc.drawLine(gc.colorDark(), rect.getTR(), rect.getBR()); 
        gc.drawLine(gc.colorHighlight(), rect.getBR(), rect.getBL()); 
        gc.drawLine(gc.colorHighlight(), rect.getBL(), rect.getTL());
    }
    
    private void draw2D (PlatformContext gc,  int c) {
        gc.drawFilledRectangle(c,rect);
    }
    

    public void update (PlatformContext gc, boolean drawGrid) {
        switch (mode) {
        case COLOR:
            draw3D(gc, color);
            break;
                
        case GREYED:
            draw3D(gc, gc.colorGrayed());
            break;
                
        case BACKGROUND:
            if (drawGrid) drawGrid(gc);
            else draw2D(gc, gc.colorBackground());
            break;
        }
    }
}

File: tlg/src/main/kotlin/tlg/Square.kt
package tlg

import tlg.context.ByteInteger
import tlg.context.PlatformContext
import tlg.geometry.TlgRectangle
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.IOException

class Square {
    private var mode = BACKGROUND
    private var color = 0
    var rect: TlgRectangle = TlgRectangle(0, 0, 0, 0)

    constructor() {}
    constructor(s: Square) {
        set(s)
    }

    constructor(reader: BufferedInputStream) {
        color = ByteInteger.read(reader)
        mode = reader.read()
    }

    @Throws(IOException::class)
    fun writeState(output: BufferedOutputStream) {
        ByteInteger.wrap(color).writeState(output)
        output.write(mode)
    }

    val isVisible: Boolean
        get() = mode != BACKGROUND
    val isInvisible: Boolean
        get() = !isVisible

    fun set(s: Square) {
        color = s.color
        mode = s.mode
    }

    fun setColor(c: Int) {
        color = c
    }

    val isActivated: Boolean
        get() = isVisible

    fun makeVisible() {
        mode = COLOR
    }

    fun makeInvisible() {
        mode = BACKGROUND
    }

    val isGreyedOut: Boolean
        get() = mode == GREYED

    fun enableGreyedOut() {
        if (isInvisible) mode = GREYED
    }

    fun disableGreyedOut() {
        if (isGreyedOut) makeInvisible()
    }

    private fun drawGrid(gc: PlatformContext) {
        gc.drawFilledRectangle(gc.colorBackground(), rect)
        gc.drawLine(gc.colorGrid(), rect.tL, rect.tR)
        gc.drawLine(gc.colorGrid(), rect.tL, rect.bL)
    }

    private fun draw3D(gc: PlatformContext, c: Int) {
        val fillRect = TlgRectangle(rect)
        fillRect.shrink(1)
        gc.drawFilledRectangle(c, fillRect)
        gc.drawLine(gc.colorDark(), rect.tL, rect.tR)
        gc.drawLine(gc.colorDark(), rect.tR, rect.bR)
        gc.drawLine(gc.colorHighlight(), rect.bR, rect.bL)
        gc.drawLine(gc.colorHighlight(), rect.bL, rect.tL)
    }

    private fun draw2D(gc: PlatformContext, c: Int) {
        gc.drawFilledRectangle(c, rect)
    }

    fun update(gc: PlatformContext, drawGrid: Boolean) {
        when (mode) {
            COLOR -> draw3D(gc, color)
            GREYED -> draw3D(gc, gc.colorGrayed())
            BACKGROUND -> if (drawGrid) drawGrid(gc) else draw2D(gc, gc.colorBackground())
        }
    }

    companion object {
        private const val BACKGROUND = 0
        private const val COLOR = 1
        private const val GREYED = 2
    }
}
